## Installing and importing a python package/library 

- Installing Pandas Library for Data Processing 

- What is Pandas? 
    - Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.
    
    - Its main data format is the DataFrame, which is like a 2D array, but with column names and indexes for the rows.

    - It allows you to work with data stored in various formats, such as CSV files, and perform tasks like cleaning, transforming, analyzing, and visualizing the data. Here's an overview of what you can do with Pandas:
        - Extracting Data: Pandas can extract data from CSV files and other formats into a DataFrame, which is a tabular structure that represents the data.

        - Calculating Statistics: You can use Pandas to calculate statistics on the data, such as finding the average, median, maximum, or minimum values for each column.
        - Analyzing Relationships: Pandas allows you to analyze relationships between columns, such as determining if there is a correlation between column A and column B.
        - Exploring Data Distribution: You can use Pandas to examine the distribution of data within a column, which can help you understand its characteristics.
        - Data Cleaning: Pandas provides functionality to clean the data, including handling missing values, filtering rows or columns based on specific criteria, and performing data transformations.
        - Data Visualization: With the help of Matplotlib, Pandas enables you to visualize the data through various types of plots such as bars, lines, histograms, and bubbles.
        - Data Storage: Once you have cleaned and transformed the data, Pandas allows you to store it back into a CSV file or other file formats, or even directly into a database.

        In summary, Pandas serves as a comprehensive tool for data management, analysis, and visualization, making it easier to extract insights and perform tasks on your data.




### Using pip to install Pandas 

In [39]:
# Need to only be run once per environment. No need to run it again in every project.

import sys
!{sys.executable} -m pip install pandas

### Importing pandas package library 

In [40]:
# Imprting pandas library as pd so instead of wrting pandas everywhere we can call pd which is shorter
import pandas as pd

## Core components of pandas: Series and DataFrames

The primary two components of pandas are the `Series` and `DataFrame`. 

A `Series` is essentially a column, and a `DataFrame` is a multi-dimensional table made up of a collection of Series. 

![Pandas Data Structure](https://www.w3resource.com/w3r_images/pandas-data-structure.svg)
![Series and Dataframe](https://storage.googleapis.com/lds-media/images/series-and-dataframe.width-1200.png)

DataFrames and Series are quite similar in that many operations that you can do with one you can do with the other, such as filling in null values and calculating the mean.

You'll see how these components work when we start working with data below. 

### Creating DataFrames from scratch
Creating DataFrames right in Python is good to know and quite useful when testing new methods and functions you find in the pandas docs.

There are *many* ways to create a DataFrame from scratch, but a great option is to just use a simple `dict`. 

Let's say we have a fruit stand that sells apples and oranges. We want to have a column for each fruit and a row for each customer purchase. To organize this as a dictionary for pandas we could do something like:

In [41]:
data = {
    'apples': [3, 2, 0, 1], 
    'oranges': [0, 3, 7, 2]
}

And then pass it to the pandas DataFrame constructor:

In [42]:
purchases = pd.DataFrame(data)

purchases

apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2

## Loading and exploring data with Pandas



First we load in the data. This particular dataset is in the **comma-separated variable (.csv)** format.

``
movies_df = pd.read_csv("IMDB-Movie-Data.csv", index_col="Title")
``

Pandas shows us the **head** (first rows) and the **tail** (last rows), as well as the **shape**. So we know there are 11 columns (separate pieces of info about each movie), and 1000 movies.


In [43]:
#load
#pd.options.display.max_rows = 100
movies_df = pd.read_csv("Datasets/IMDB/IMDB-Movie-Data.csv", index_col="Title")

In [44]:
# Standard format is rows x colums
# Hence our dataset has 1000 rows and 11 colums
movies_df.shape

(1000, 11)

In [45]:
# Getting a list of all column names in my dataframe
movies_df.keys()

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

### Viewing your data

The first thing to do when opening a new dataset is print out a few rows using `.head()`:

In [46]:
movies_df.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

`.head()` outputs the **first** five rows of your DataFrame by default, but we could also pass a number as well: `movies_df.head(10)` would output the top ten rows, for example. 

To see the **last** five rows use `.tail()`. `tail()` also accepts a number, and in this case we printing the bottom two rows.:

In [47]:
movies_df.tail(2)

Rank                  Genre  \
Title                                       
Search Party   999       Adventure,Comedy   
Nine Lives    1000  Comedy,Family,Fantasy   

                                                    Description  \
Title                                                             
Search Party  A pair of friends embark on a mission to reuni...   
Nine Lives    A stuffy businessman finds himself trapped ins...   

                      Director  \
Title                            
Search Party    Scot Armstrong   
Nine Lives    Barry Sonnenfeld   

                                                         Actors  Year  \
Title                                                                   
Search Party  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
Nine Lives    Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

              Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Title                                                                          
Search Party                 93     5.6   4881                 NaN       22.0  
Nine Lives                   87     5.3  12435               19.64       11.0

Typically when we load in a dataset, we like to view the first five or so rows to see what's under the hood. Here we can see the names of each column, the index, and examples of values in each row.

You'll notice that the index in our DataFrame is the *Title* column, which you can tell by how the word *Title* is slightly lower than the rest of the columns.

### Getting info about your data

`.info()` should be one of the very first commands you run after loading your data:

In [48]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


`.info()` provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using. 

Notice in our movies dataset we have some obvious missing values in the `Revenue` and `Metascore` columns. We'll look at how to handle those in a bit.

Seeing the datatype quickly is actually quite useful. Imagine you just imported some JSON and the integers were recorded as strings. You go to do some arithmetic and find an "unsupported operand" Exception because you can't do math with strings. Calling `.info()` will quickly point out that your column you thought was all integers are actually string objects.

In [49]:
# What are the types?
# Checking the data type of each columne

movies_df.dtypes

Rank                    int64
Genre                  object
Description            object
Director               object
Actors                 object
Year                    int64
Runtime (Minutes)       int64
Rating                float64
Votes                   int64
Revenue (Millions)    float64
Metascore             float64
dtype: object

### Handling duplicates

This dataset does not have duplicate rows, but it is always important to verify you aren't aggregating duplicate rows. 

To demonstrate, let's simply just double up our movies DataFrame by appending it to itself:

In [50]:
# .append is now depricated in the new pandas version. Use concat instead 
# temp_df = movies_df.append(movies_df)

# temp_df.shape

temp_df = pd.concat([movies_df, movies_df])
temp_df.shape

(2000, 11)

Using `append()` will return a copy without affecting the original DataFrame. We are capturing this copy in `temp` so we aren't working with the real data.

Notice call `.shape` quickly proves our DataFrame rows have doubled.

Now we can try dropping duplicates:

In [51]:
temp_df = temp_df.drop_duplicates()

temp_df.shape

(1000, 11)

Just like `append()`, the `drop_duplicates()` method will also return a copy of your DataFrame, but this time with duplicates removed. Calling `.shape` confirms we're back to the 1000 rows of our original dataset.

To avoid assigning DataFrames to the same variable multiple times, you can use the `inplace=True` option in pandas. This will modify the DataFrame object directly.

In [52]:
temp_df.drop_duplicates(inplace=True)

Now our `temp_df` *will* have the transformed data automatically. 

Another important argument for `drop_duplicates()` is `keep`, which has three possible options:

* `first`: (default) Drop duplicates except for the first occurrence.
* `last`: Drop duplicates except for the last occurrence.
* `False`: Drop all duplicates.

Since we didn't define the `keep` arugment in the previous example it was defaulted to `first`. This means that if two rows are the same pandas will drop the second row and keep the first row. Using `last` has the opposite effect: the first row is dropped.

`keep`, on the other hand, will drop all duplicates. If two rows are the same then both will be dropped. Watch what happens to `temp_df`:

In [53]:
# temp_df = movies_df.concat(movies_df)  # make a new copy

temp_df = pd.concat([movies_df, movies_df])

temp_df.drop_duplicates(inplace=True, keep=False)

temp_df.shape

(0, 11)

Since all rows were duplicates, `keep=False` dropped them all resulting in zero rows being left over. If you're wondering why you would want to do this, one reason is that it allows you to locate all duplicates in your dataset. When conditional selections are shown below you'll see how to do that.

### Column cleanup

Many times datasets will have verbose column names with symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we can spend a little time cleaning up their names.

Here's how to print the column names of our dataset:

In [54]:
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

Not only does `.columns` come in handy if you want to rename columns by allowing for simple copy and paste, it's also useful if you need to understand why you are receiving a `Key Error` when selecting data by column.

We can use the `.rename()` method to rename certain or all columns via a `dict`. We don't want parentheses, so let's rename those:

In [55]:
movies_df.rename(columns={
        'Runtime (Minutes)': 'Runtime', 
        'Revenue (Millions)': 'Revenue_millions'
    }, inplace=True)


movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue_millions', 'Metascore'],
      dtype='object')

- Pass the list of new column names to df.columns. Now, the new column names will be assigned to the DataFrame. 
- For this to work, you need to pass the names of all the columns. 
- The length of this names list must be exactly equal to the total number of columns in the DataFrame.

In [56]:
movies_df.columns = ['test', 'genre', 'description', 'director', 'actors', 'year', 'runtime', 
                     'rating', 'votes', 'revenue_millions', 'metascore']


movies_df.columns

Index(['test', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

In [57]:
movies_df.columns = ['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime', 
                     'rating', 'votes', 'revenue_millions', 'metascore']


movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

### How to work with missing values

When exploring data, you’ll most likely encounter missing or null values, which are essentially placeholders for non-existent values. Most commonly you'll see Python's `None` or NumPy's `np.nan`, each of which are handled differently in some situations.

There are two options in dealing with nulls: 

1. Get rid of rows or columns with nulls
2. Replace nulls with non-null values, a technique known as **imputation**

Let's calculate to total number of nulls in each column of our dataset. The first step is to check which cells in our DataFrame are null:

In [58]:
movies_df.isnull()

rank  genre  description  director  actors   year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         runtime  rating  votes  revenue_millions  metascore  
Title                                                                         
Guardians of the Galaxy    False   False  False             False      False  
Prometheus                 False   False  False             False      False  
Split                      False   False  False             False      False  
Sing                       False   False  False             False      False  
Suicide Squad              False   False  False             False      False  
...                          ...     ...    ...               ...        ...  
Secret in Their Eyes       False   False  False              True      False  
Hostel: Part II            False   False  False             False      False  
Step Up 2: The Streets     False   False  False             False      False  
Search Party               False   False  False              True      False  
Nine Lives                 False   False  False             False      False  

[1000 rows x 11 columns]

Notice `isnull()` returns a DataFrame where each cell is either True or False depending on that cell's null status.

To count the number of nulls in each column we use an aggregate function for summing: 

In [59]:
movies_df.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

`.isnull()` just by iteself isn't very useful, and is usually used in conjunction with other methods, like `sum()`.

We can see now that our data has **128** missing values for `revenue_millions` and **64** missing values for `metascore`.

In [93]:
# Bonus Trick -- If you wanted get a sesnse of missing value in percentages 
(movies_df.isnull().sum()/len(movies_df))*100

rank                0.0
genre               0.0
description         0.0
director            0.0
actors              0.0
year                0.0
runtime             0.0
rating              0.0
votes               0.0
revenue_millions    0.0
metascore           6.4
dtype: float64

#### Removing null values

Data Scientists and Analysts regularly face the dilemma of dropping or imputing null values, and is a decision that requires intimate knowledge of your data and its context. Overall, removing null data is only suggested if you have a small amount of missing data.

Remove nulls is pretty simple:

In [60]:
movies_df

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         year  runtime  rating   votes  revenue_millions  \
Title                                                                      
Guardians of the Galaxy  2014      121     8.1  757074            333.13   
Prometheus               2012      124     7.0  485820            126.46   
Split                    2016      117     7.3  157606            138.12   
Sing                     2016      108     7.2   60545            270.32   
Suicide Squad            2016      123     6.2  393727            325.02   
...                       ...      ...     ...     ...   

In [61]:
movies_df.shape

(1000, 11)

In [62]:
movies_df.dropna()

rank                     genre  \
Title                                                      
Guardians of the Galaxy      1   Action,Adventure,Sci-Fi   
Prometheus                   2  Adventure,Mystery,Sci-Fi   
Split                        3           Horror,Thriller   
Sing                         4   Animation,Comedy,Family   
Suicide Squad                5  Action,Adventure,Fantasy   
...                        ...                       ...   
Resident Evil: Afterlife   994   Action,Adventure,Horror   
Project X                  995                    Comedy   
Hostel: Part II            997                    Horror   
Step Up 2: The Streets     998       Drama,Music,Romance   
Nine Lives                1000     Comedy,Family,Fantasy   

                                                                description  \
Title                                                                         
Guardians of the Galaxy   A group of intergalactic criminals are forced ...   
Prometheus                Following clues to the origin of mankind, a te...   
Split                     Three girls are kidnapped by a man with a diag...   
Sing                      In a city of humanoid animals, a hustling thea...   
Suicide Squad             A secret government agency recruits some of th...   
...                                                                     ...   
Resident Evil: Afterlife  While still out to destroy the evil Umbrella C...   
Project X                 3 high school seniors throw a birthday party t...   
Hostel: Part II           Three American college students studying abroa...   
Step Up 2: The Streets    Romantic sparks occur between two dance studen...   
Nine Lives                A stuffy businessman finds himself trapped ins...   

                                      director  \
Title                                            
Guardians of the Galaxy             James Gunn   
Prometheus                        Ridley Scott   
Split                       M. Night Shyamalan   
Sing                      Christophe Lourdelet   
Suicide Squad                       David Ayer   
...                                        ...   
Resident Evil: Afterlife    Paul W.S. Anderson   
Project X                      Nima Nourizadeh   
Hostel: Part II                       Eli Roth   
Step Up 2: The Streets              Jon M. Chu   
Nine Lives                    Barry Sonnenfeld   

                                                                     actors  \
Title                                                                         
Guardians of the Galaxy   Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus                Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                     James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                      Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad             Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                     ...   
Resident Evil: Afterlife  Milla Jovovich, Ali Larter, Wentworth Miller,K...   
Project X                 Thomas Mann, Oliver Cooper, Jonathan Daniel Br...   
Hostel: Part II           Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets    Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Nine Lives                Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                          year  runtime  rating   votes  revenue_millions  \
Title                                                                       
Guardians of the Galaxy   2014      121     8.1  757074            333.13   
Prometheus                2012      124     7.0  485820            126.46   
Split                     2016      117     7.3  157606            138.12   
Sing                      2016      108     7.2   60545            270.32   
Suicide Squad             2016      123     6.2  393727            325.02   

This operation will delete any **row** with at least a single null value, but it will return a new DataFrame without altering the original one. You could specify `inplace=True` in this method as well.

So in the case of our dataset, this operation would remove 128 rows where `revenue_millions` is null and 64 rows where `metascore` is null. This obviously seems like a waste since there's perfectly good data in the other columns of those dropped rows. That's why we'll look at imputation next.

Other than just dropping rows, you can also drop columns with null values by setting `axis=1`:

In [63]:
movies_df.dropna(axis=1)

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         year  runtime  rating   votes  
Title                                                   
Guardians of the Galaxy  2014      121     8.1  757074  
Prometheus               2012      124     7.0  485820  
Split                    2016      117     7.3  157606  
Sing                     2016      108     7.2   60545  
Suicide Squad            2016      123     6.2  393727  
...                       ...      ...     ...     ...  
Secret in Their Eyes     2015      111     6.2   27585  
Hostel: Part II          2007       94     5.5   73152  
Step Up 2: The Stre

In our dataset, this operation would drop the `revenue_millions` and `metascore` columns.

**Intuition side note**: What's with this `axis=1` parameter?

It's not immediately obvious where `axis` comes from and why you need it to be 1 for it to affect columns. To see why, just look at the `.shape` output:

In [64]:
movies_df.shape
# Rows are at index zero and colums are at index one (1000, 11)

(1000, 11)

### Imputation

Imputation is a conventional feature engineering technique used to keep valuable data that have null values. 

There may be instances where dropping every row with a null value removes too big a chunk from your dataset, so instead we can impute that null with another value, usually the **mean** or the **median** of that column. 

Let's look at imputing the missing values in the `revenue_millions` column. First we'll extract that column into its own variable:

In [65]:
revenue = movies_df['revenue_millions']

Using square brackets is the general way we select columns in a DataFrame. 

If you remember back to when we created DataFrames from scratch, the keys of the `dict` ended up as column names. Now when we select columns of a DataFrame, we use brackets just like if we were accessing a Python dictionary. 

`revenue` now contains a Series:

In [66]:
revenue.head()

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
Name: revenue_millions, dtype: float64

Slightly different formatting than a DataFrame, but we still have our `Title` index. 

We'll impute the missing values of revenue using the mean. Here's the mean value:

In [67]:
revenue_mean = revenue.mean()

revenue_mean

np.float64(82.95637614678898)

With the mean, let's fill the nulls using `fillna()`:

In [68]:
revenue.fillna(revenue_mean, inplace=True)

We have now replaced all nulls in `revenue` with the mean of the column. Notice that by using `inplace=True` we have actually affected the original `movies_df`:

In [69]:
movies_df.isnull().sum()

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

In [70]:
movies_df

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         year  runtime  rating   votes  revenue_millions  \
Title                                                                      
Guardians of the Galaxy  2014      121     8.1  757074        333.130000   
Prometheus               2012      124     7.0  485820        126.460000   
Split                    2016      117     7.3  157606        138.120000   
Sing                     2016      108     7.2   60545        270.320000   
Suicide Squad            2016      123     6.2  393727        325.020000   
...                       ...      ...     ...     ...   

### Summarising Data

**Pandas** can also give us a summary of our data (1000 is a lot to look at ourselves!). 

``movies_df.describe(include = "all")``


movies_df.describe()

If we don't put ``include = "all"``, then we will only get summary stastics for **numeric** columns. 


In [71]:
movies_df.describe(include = "all")

rank                    genre  \
count   1000.000000                     1000   
unique          NaN                      207   
top             NaN  Action,Adventure,Sci-Fi   
freq            NaN                       50   
mean     500.500000                      NaN   
std      288.819436                      NaN   
min        1.000000                      NaN   
25%      250.750000                      NaN   
50%      500.500000                      NaN   
75%      750.250000                      NaN   
max     1000.000000                      NaN   

                                              description      director  \
count                                                1000          1000   
unique                                               1000           644   
top     A group of intergalactic criminals are forced ...  Ridley Scott   
freq                                                    1             8   
mean                                                  NaN           NaN   
std                                                   NaN           NaN   
min                                                   NaN           NaN   
25%                                                   NaN           NaN   
50%                                                   NaN           NaN   
75%                                                   NaN           NaN   
max                                                   NaN           NaN   

                                                   actors         year  \
count                                                1000  1000.000000   
unique                                                996          NaN   
top     Jennifer Lawrence, Josh Hutcherson, Liam Hemsw...          NaN   
freq                                                    2          NaN   
mean                                                  NaN  2012.783000   
std                                                   NaN     3.205962   
min                                                   NaN  2006.000000   
25%                                                   NaN  2010.000000   
50%                                                   NaN  2014.000000   
75%                                                   NaN  2016.000000   
max                                                   NaN  2016.000000   

            runtime       rating         votes  revenue_millions   metascore  
count   1000.000000  1000.000000  1.000000e+03       1000.000000  936.000000  
unique          NaN          NaN           NaN               NaN         NaN  
top             NaN          NaN           NaN               NaN         NaN  
freq            NaN          NaN           NaN               NaN         NaN  
mean     113.172000     6.723200  1.698083e+05         82.956376   58.985043  
std       18.810908     0.945429  1.887626e+05         96.412043   17.194757  
min       66.000000     1.900000  6.100000e+01          0.000000   11.000000  
25%      100.000000     6.200000  3.630900e+04         17.442500   47.000000  
50%      111.000000     6.800000  1.107990e+05         60.375000   59.500000  
75%      123.000000     7.400000  2.399098e+05         99.177500   72.000000  
max      191.000000     9.000000  1.791916e+06        936.630000  100.000000

For example his tells us that the genre column has 207 unique values, the top value is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

`.value_counts()` can tell us the frequency of all values in a column:

In [72]:
movies_df['genre'].value_counts().head(10)

genre
Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Animation,Adventure,Comedy    27
Action,Adventure,Fantasy      27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: count, dtype: int64

### DataFrame slicing, selecting, extracting

Up until now we've focused on some basic summaries of our data. We've learned about simple column extraction using single brackets, and we imputed null values in a column using `fillna()`. Below are the other methods of slicing, selecting, and extracting you'll need to use constantly.

It's important to note that, although many methods are the same, DataFrames and Series have different attributes, so you'll need be sure to know which type you are working with or else you will receive attribute errors. 

Let's look at working with columns first.

#### By column

We can select a column using its name 

``df["ColumnName"]``

Or we can select a bunch of columns by passing an array 

``df[["ColumnName1","ColumnName2"]]``

This returns a smaller **Series** object with the results but we can also use 

``result.values``


In [73]:
# You already saw how to extract a column using square brackets like this:

genre_col = movies_df['genre']
print(genre_col)

Title
Guardians of the Galaxy     Action,Adventure,Sci-Fi
Prometheus                 Adventure,Mystery,Sci-Fi
Split                               Horror,Thriller
Sing                        Animation,Comedy,Family
Suicide Squad              Action,Adventure,Fantasy
                                     ...           
Secret in Their Eyes            Crime,Drama,Mystery
Hostel: Part II                              Horror
Step Up 2: The Streets          Drama,Music,Romance
Search Party                       Adventure,Comedy
Nine Lives                    Comedy,Family,Fantasy
Name: genre, Length: 1000, dtype: object


In [74]:
type(genre_col)

pandas.core.series.Series

In [75]:
# To extract a column as a DataFrame, use double square brackets
genre_col = movies_df[['genre']]

In [76]:
type(genre_col)

pandas.core.frame.DataFrame

In [77]:
# We can also have multiple columns
subset = movies_df[['genre', 'rating']]

subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

In [78]:
#select column and get it back as an array of values
movies_df["rating"].values

array([8.1, 7. , 7.3, 7.2, 6.2, 6.1, 8.3, 6.4, 7.1, 7. , 7.5, 7.8, 7.9,
       7.7, 6.4, 6.6, 8.2, 6.7, 8.1, 8. , 6.7, 7.9, 6.7, 6.5, 5.3, 6.8,
       8.3, 4.7, 6.2, 5.9, 6.3, 7.5, 7.1, 8. , 5.6, 7.9, 8.6, 7.6, 6.9,
       7.1, 6.3, 7.5, 2.7, 7.2, 6.3, 6.7, 7.3, 5.6, 7.1, 3.7, 8.1, 5.8,
       5.6, 7.2, 9. , 7.3, 7.2, 7.4, 7. , 7.5, 6.7, 6.8, 6.5, 4.1, 8.5,
       7.7, 7.4, 8.1, 7.5, 7.2, 5.9, 7.1, 7.5, 6.8, 8.1, 7.1, 8.1, 8.3,
       7.3, 5.3, 8.8, 7.9, 8.2, 8.1, 7.2, 7. , 6.4, 7.8, 7.8, 7.4, 8.1,
       7. , 8.1, 7.1, 7.4, 7.4, 8.6, 5.8, 6.3, 8.5, 7. , 7. , 8. , 7.9,
       7.3, 7.7, 5.4, 6.3, 5.8, 7.7, 6.3, 8.1, 6.1, 7.7, 8.1, 5.8, 6.2,
       8.8, 7.2, 7.4, 6.7, 6.7, 6. , 7.4, 8.5, 7.5, 5.7, 6.6, 6.4, 8. ,
       7.3, 6. , 6.4, 8.5, 7.1, 7.3, 8.1, 7.3, 8.1, 7.1, 8. , 6.2, 7.8,
       8.2, 8.4, 8.1, 7.4, 7.6, 7.6, 6.2, 6.4, 7.2, 5.8, 7.6, 8.1, 4.7,
       7. , 7.4, 7.5, 7.9, 6. , 7. , 8. , 6.1, 8. , 5.2, 6.5, 7.3, 7.3,
       6.8, 7.9, 7.9, 5.2, 8. , 7.5, 6.5, 7.6, 7. , 7.4, 7.3, 6.

For more information on the differnce between lists and arrays check out https://www.geeksforgeeks.org/difference-between-list-and-array-in-python/

#### By rows

For rows, we have two options: 

- `.loc` - **loc**ates by name
- `.iloc`- **loc**ates by numerical **i**ndex

Remember that we are still indexed by movie Title, so to use `.loc` we give it the Title of a movie:

In [79]:
prom = movies_df.loc["Prometheus"]

prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                            7.0
votes                                                          485820
revenue_millions                                               126.46
metascore                                                        65.0
Name: Prometheus, dtype: object

On the other hand, with `iloc` we give it the numerical index of Prometheus:

In [80]:
prom = movies_df.iloc[1]

In [81]:
prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                            7.0
votes                                                          485820
revenue_millions                                               126.46
metascore                                                        65.0
Name: Prometheus, dtype: object

`loc` and `iloc` can be thought of as similar to Python `list` slicing. To show this even further, let's select multiple rows.

How would you do it with a list? In Python, just slice with brackets like `example_list[1:4]`. It's works the same way in pandas:

In [82]:
movie_subset = movies_df.loc['Prometheus':'Sing']

movie_subset = movies_df.iloc[1:4]

movie_subset

rank                     genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       actors  year  runtime  \
Title                                                                          
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012      124   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016      117   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016      108   

            rating   votes  revenue_millions  metascore  
Title                                                    
Prometheus     7.0  485820            126.46       65.0  
Split          7.3  157606            138.12       62.0  
Sing           7.2   60545            270.32       59.0

One important distinction between using `.loc` and `.iloc` to select multiple rows is that `.loc` includes the movie *Sing* in the result, but when using `.iloc` we're getting rows 1:4 but the movie at index 4 (*Suicide Squad*) is not included. 

Slicing with `.iloc` follows the same rules as slicing with lists, the object at the index at the end is not included.



In [83]:
movies_df.iloc[1:4, 2:4]

description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  
Title                             
Prometheus          Ridley Scott  
Split         M. Night Shyamalan  
Sing        Christophe Lourdelet

The expression movies_df.iloc[1:4, 2:4] shows only columns 2 and 3 (not the 4th column) because in Python's slice notation:
- The second number in a slice is exclusive (It means the ending number you specify is not included in the result. It stops right before that number.)
- Indexing starts at 0

2:4 means:
- Start at index 2 (third column)
- Stop before index 4 (don't include the fourth column)
- Only includes columns 2 and 3

#### Counting Columns

We can get counts to see what the most prevalent combinations of categories are. 



In [84]:
movies_df.keys()

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

In [85]:
movies_df[["genre","director"]].value_counts()

genre                     director       
Adventure,Family,Fantasy  David Yates        4
Action,Adventure,Sci-Fi   Michael Bay        4
Action,Crime,Thriller     Justin Lin         4
Drama                     Lars von Trier     3
Adventure,Fantasy         Peter Jackson      3
                                            ..
Adventure,Horror          Jackson Stewart    1
Adventure,Horror,Mystery  Christophe Gans    1
Adventure,Mystery,Sci-Fi  Neil Burger        1
                          Ridley Scott       1
Thriller,War              Fabio Guaglione    1
Name: count, Length: 950, dtype: int64

#### Conditional selections
We’ve gone over how to select columns and rows, but what if we want to make a conditional selection? 

For example, what if we want to filter our movies DataFrame to show only films directed by Ridley Scott or films with a rating greater than or equal to 8.0?

To do that, we take a column from the DataFrame and apply a Boolean condition to it. Here's an example of a Boolean condition:

In [86]:
condition = (movies_df['director'] == "Ridley Scott")

condition.head()

Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
Name: director, dtype: bool

Similar to `isnull()`, this returns a Series of True and False values: True for films directed by Ridley Scott and False for ones not directed by him. 

We want to filter out all movies not directed by Ridley Scott, in other words, we don’t want the False films. To return the rows where that condition is True we have to pass this operation into the DataFrame:

In [87]:
movies_df[movies_df['director'] == "Ridley Scott"].head()

rank                     genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   

                                                              description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
The Martian             An astronaut becomes stranded on Mars after hi...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   

                            director  \
Title                                  
Prometheus              Ridley Scott   
The Martian             Ridley Scott   
Robin Hood              Ridley Scott   
American Gangster       Ridley Scott   
Exodus: Gods and Kings  Ridley Scott   

                                                                   actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   

                        year  runtime  rating   votes  revenue_millions  \
Title                                                                     
Prometheus              2012      124     7.0  485820            126.46   
The Martian             2015      144     8.0  556097            228.43   
Robin Hood              2010      140     6.7  221117            105.22   
American Gangster       2007      157     7.8  337835            130.13   
Exodus: Gods and Kings  2014      150     6.0  137299             65.01   

                        metascore  
Title                              
Prometheus                   65.0  
The Martian                  80.0  
Robin Hood                   53.0  
American Gangster            76.0  
Exodus: Gods and Kings       52.0

You can get used to looking at these conditionals by reading it like: 

> Select movies_df where movies_df director equals Ridley Scott

Let's look at conditional selections using numerical values by filtering the DataFrame by ratings:

In [88]:
movies_df[movies_df['rating'] >= 8.6].head(3)

rank                    genre  \
Title                                            
Interstellar       37   Adventure,Drama,Sci-Fi   
The Dark Knight    55       Action,Crime,Drama   
Inception          81  Action,Adventure,Sci-Fi   

                                                       description  \
Title                                                                
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
Inception        A thief, who steals corporate secrets through ...   

                          director  \
Title                                
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
Inception        Christopher Nolan   

                                                            actors  year  \
Title                                                                      
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 runtime  rating    votes  revenue_millions  metascore  
Title                                                                   
Interstellar         169     8.6  1047747            187.99       74.0  
The Dark Knight      152     9.0  1791916            533.32       82.0  
Inception            148     8.8  1583625            292.57       74.0

We can make some richer conditionals by using logical operators `|` for "or" and `&` for "and".

Let's filter the the DataFrame to show only movies by Christopher Nolan OR Ridley Scott:

In [94]:
movies_df[(movies_df['director'] == 'Christopher Nolan') | (movies_df['director'] == 'Ridley Scott')].head()

rank                     genre  \
Title                                             
Prometheus          2  Adventure,Mystery,Sci-Fi   
Interstellar       37    Adventure,Drama,Sci-Fi   
The Dark Knight    55        Action,Crime,Drama   
The Prestige       65      Drama,Mystery,Sci-Fi   
Inception          81   Action,Adventure,Sci-Fi   

                                                       description  \
Title                                                                
Prometheus       Following clues to the origin of mankind, a te...   
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
The Prestige     Two stage magicians engage in competitive one-...   
Inception        A thief, who steals corporate secrets through ...   

                          director  \
Title                                
Prometheus            Ridley Scott   
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
The Prestige     Christopher Nolan   
Inception        Christopher Nolan   

                                                            actors  year  \
Title                                                                      
Prometheus       Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
The Prestige     Christian Bale, Hugh Jackman, Scarlett Johanss...  2006   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 runtime  rating    votes  revenue_millions  metascore  
Title                                                                   
Prometheus           124     7.0   485820            126.46       65.0  
Interstellar         169     8.6  1047747            187.99       74.0  
The Dark Knight      152     9.0  1791916            533.32       82.0  
The Prestige         130     8.5   913152             53.08       66.0  
Inception            148     8.8  1583625            292.57       74.0

### Sorting

We can do that using ``sort_values``. We tell **Pandas** which column we want to sort by, and this must be either a number, or an **ordinal** value (such as a date). We also say which direction we want to sort the results in, and this is useful if we want to get the top or bottom slice

In [90]:
movies_df.sort_values(by='rank', ascending=False)[:5]

rank                  genre  \
Title                                                 
Nine Lives              1000  Comedy,Family,Fantasy   
Search Party             999       Adventure,Comedy   
Step Up 2: The Streets   998    Drama,Music,Romance   
Hostel: Part II          997                 Horror   
Secret in Their Eyes     996    Crime,Drama,Mystery   

                                                              description  \
Title                                                                       
Nine Lives              A stuffy businessman finds himself trapped ins...   
Search Party            A pair of friends embark on a mission to reuni...   
Step Up 2: The Streets  Romantic sparks occur between two dance studen...   
Hostel: Part II         Three American college students studying abroa...   
Secret in Their Eyes    A tight-knit team of rising investigators, alo...   

                                director  \
Title                                      
Nine Lives              Barry Sonnenfeld   
Search Party              Scot Armstrong   
Step Up 2: The Streets        Jon M. Chu   
Hostel: Part II                 Eli Roth   
Secret in Their Eyes           Billy Ray   

                                                                   actors  \
Title                                                                       
Nine Lives              Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   
Search Party            Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Step Up 2: The Streets  Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Hostel: Part II         Lauren German, Heather Matarazzo, Bijou Philli...   
Secret in Their Eyes    Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   

                        year  runtime  rating  votes  revenue_millions  \
Title                                                                    
Nine Lives              2016       87     5.3  12435         19.640000   
Search Party            2014       93     5.6   4881         82.956376   
Step Up 2: The Streets  2008       98     6.2  70699         58.010000   
Hostel: Part II         2007       94     5.5  73152         17.540000   
Secret in Their Eyes    2015      111     6.2  27585         82.956376   

                        metascore  
Title                              
Nine Lives                   11.0  
Search Party                 22.0  
Step Up 2: The Streets       50.0  
Hostel: Part II              46.0  
Secret in Their Eyes         45.0